design-documents/frontend/advanced-js-bundling/advanced-js-bundling.md

```diff
@@ -0,0 +1,82 @@
+-   **Bundle**: One or more JavaScript files (created by `bundlegento` automatically) that are a collection of > 1 module
+
+## Context
+
+JavaScript module bundling is not a new idea, and Magento is far from the first piece of software to implement it. There are many excellent off-the-shelf tools designed to solve this problem:
+
+-   [Browserify](http://browserify.org/)
+-   [webpack](https://webpack.js.org/)
+-   [Parcel](https://parceljs.org/)
+
+The problem, though, is that these tools were designed for applications where the entirety of the graph can be discovered just through static analysis of JavaScript source files.
+
+This is not the case for Magento 2. Although Magento does use a proper module system where each module can express its dependencies (RequireJS), it is impossible to statically analyze which entry points will be used for an individual page.
+
+This is because RequireJS supports fetching/loading of arbitrary modules at runtime, through the global variables require and define. Unfortunately, it has become a common practice (and in some places, a documented feature) in Magento to rely on heavily dynamic use of require/define.
+
+Sources of Dynamic Module Entries in Magento 2
+
+-   [`x-magento-init`/`mage-init`](https://devdocs.magento.com/guides/v2.3/javascript-dev-guide/javascript/js_init.html)
+-   [UI Components](https://devdocs.magento.com/guides/v2.3/ui_comp_guide/bk-ui_comps.html) (`template` property)
+-   [Inline calls](https://github.com/magento/magento2/blob/07e57ac249e02b721613172c4c3c8e24a3bdd746/app/code/Magento/Sales/view/adminhtml/templates/order/create/form/address.phtml#L32)
+-   [Mixins](https://devdocs.magento.com/guides/v2.3/javascript-dev-guide/javascript/js_mixins.html)
+
+Also our [existing JS bundling](https://devdocs.magento.com/guides/v2.3/frontend-dev-guide/themes/js-bundling.html) merge *all* files into bundles which brings bigger size of static content (2x time bigger)
+
+## Problems
+
+We have 180 HTTP requests to fetch all needed JS code just for empty Home Page and overall JS size is 2.2 MB. According to google - mobile user with slow connection can have up to 8sec main-thread work time, bad pagespeed insight score, and over 21sec time to interactive.
+
+Better JS bundling approach can reduce package size by several times. This is one of the most valuable enhancement can be done from storefront side performance perspective. According to our measurements, advanced bundling can improve overall performance score, main-thread work time and TTI up to 2-3x time.
+
+## Solution
+
+I propose to introduce 'Bundlegento' as separate tool for parsing Magento store pages and generate page specific bundles.
+Currently we have prototype written in NodeJS that parse HTML pages and can generate needed bundles for Magento store. Since that this is not version specific solution, it will support all Magento released versions
```

Comment:
The only pitfall I can see is the possibility that the development and production environments may be in totally different states. This may lead to the situation when module is requested on the former but is not on the latter or otherwise. This shouldn't be a big problem though as the potential overhead should not be to big and requirejs should be able to load any missing modules at runtime as usual.